File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        if(s.length() < 1){
            return 0;
        }
        
        //2 pointers 
        int left = 0;
        int right = 0;
        int max = Integer.MIN_VALUE;
        
        HashSet<Character> set = new HashSet<>();
        
        while(right < s.length()){
            
            if(!set.contains(s.charAt(right))){
                set.add(s.charAt(right));
                right++;
                max = Math.max(max,set.size());
            } else{
                set.remove(s.charAt(left));
                left++;
            }
        }
        
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
        let left = 0;
        let right = 0;
        let max = 0;

        const set = new Set();

        while (right < s.length) {
            if (!set.has(s.charAt(right))) {
                set.add(s.charAt(right));
                max = Math.max(max, right - left + 1);
                right++;
            } else {
                set.delete(s.charAt(left));
                left++;
            }
        }

        return max;
    };